import into failed with error "lock held by current process" after injection pd leader network partition

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
import into failed after simulated pd leader io delay 1s


import into failed when simulated pd io delay
```
./tidb-server -V
Release Version: v8.5.5
Edition: Community
Git Commit Hash: 0c53c76
Git Branch: HEAD
UTC Build Time: 2025-12-25 08:52:50
GoVersion: go1.25.5
Race Enabled: false
Check Table Before Drop: false
Store: unistore
2025-12-29T12:07:12.638+0800
```
operatorLogs:
```
[2025-11-29 22:06:15] ###### start import into
import into user_data1 from 's3://qe-testing/brie/lightning-mhy-100G-csv/sysbench.user_data1.000001*.csv?access-key=xxxxxx&secret-access-key=xxxxxx&endpoint=http%3a%2f%2fks3-cn-beijing-internal.ksyuncs.com&force-path-style=false&region=Beijing&provider=ks' WITH DETACHED,thread=8,skip_rows=1
[2025-11-29 22:06:15] ###### wait for import job to finish
[2025-11-29 22:24:46] ###### wait for import job to finish failed
select id, table_schema, table_name, step, status,parameters from mysql.tidb_import_jobs where start_time >= '2025-11-29 22:06:15' and table_schema = 'importtest' and table_name = 'user_data1'
jobId: 120002, table_schema: importtest, table_name: user_data1, step: importing, status: failed
```
```
./tidb-server -V
 Release Version: v9.0.0-beta.2.pre-834-g5bcc229
Edition: Community
Git Commit Hash: 5bcc2299e457bb44d8d7a769f7a4162b310605b6
Git Branch: HEAD
UTC Build Time: 2025-11-28 10:30:24
GoVersion: go1.23.12
Race Enabled: false
Check Table Before Drop: false
Store: unistore
Kernel Type: Classic
2025-11-29T22:06:00.991+0800
```
```
[2025/11/29 22:24:19.012 +08:00] [INFO] [task_executor.go:508] ["run subtask failed"] [task-id=120002] [task-key=ImportInto/120002] [subtaskID=120006] [step=import] [takeTime=15.765984ms] [error="lock held by current process"] [errorVerbose="lock held by current process\n(1) attached stack trace\n  -- stack trace:\n  | github.com/cockroachdb/pebble/vfs.defaultFS.Lock\n  | \t/root/go/pkg/mod/github.com/cockroachdb/pebble@v1.1.4-0.20250120151818-5dd133a1e6fb/vfs/file_lock_unix.go:50\n  | github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).Lock\n  | \t/root/go/pkg/mod/github.com/cockroachdb/pebble@v1.1.4-0.20250120151818-5dd133a1e6fb/vfs/disk_health.go:679\n  | github.com/cockroachdb/pebble.LockDirectory\n  | \t/root/go/pkg/mod/github.com/cockroachdb/pebble@v1.1.4-0.20250120151818-5dd133a1e6fb/open.go:1068\n  | github.com/cockroachdb/pebble.Open\n  | \t/root/go/pkg/mod/github.com/cockroachdb/pebble@v1.1.4-0.20250120151818-5dd133a1e6fb/open.go:116\n  | github.com/pingcap/tidb/pkg/lightning/backend/local.(*engineManager).openEngineDB\n  | \t/workspace/source/tidb/pkg/lightning/backend/local/engine_mgr.go:229\n  | github.com/pingcap/tidb/pkg/lightning/backend/local.(*engineManager).openEngine\n  | \t/workspace/source/tidb/pkg/lightning/backend/local/engine_mgr.go:235\n  | github.com/pingcap/tidb/pkg/lightning/backend/local.(*Backend).OpenEngine\n  | \t/workspace/source/tidb/pkg/lightning/backend/local/local.go:833\n  | github.com/pingcap/tidb/pkg/lightning/backend.EngineManager.OpenEngine\n  | \t/workspace/source/tidb/pkg/lightning/backend/backend.go:270\n  | github.com/pingcap/tidb/pkg/executor/importer.(*TableImporter).OpenIndexEngine\n  | \t/workspace/source/tidb/pkg/executor/importer/table_import.go:513\n  | github.com/pingcap/tidb/pkg/disttask/importinto.(*importStepExecutor).RunSubtask\n  | \t/workspace/source/tidb/pkg/disttask/importinto/task_executor.go:237\n  | github.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.(*BaseTaskExecutor).runSubtask.func1\n  | \t/workspace/source/tidb/pkg/disttask/framework/taskexecutor/task_executor.go:504\n  | github.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.(*BaseTaskExecutor).runSubtask\n  | \t/workspace/source/tidb/pkg/disttask/framework/taskexecutor/task_executor.go:505\n  | github.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.(*BaseTaskExecutor).Run\n  | \t/workspace/source/tidb/pkg/disttask/framework/taskexecutor/task_executor.go:379\n  | github.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.(*Manager).startTaskExecutor.func2\n  | \t/workspace/source/tidb/pkg/disttask/framework/taskexecutor/manager.go:357\n  | github.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).RunWithLog.func1\n  | \t/workspace/source/tidb/pkg/util/wait_group_wrapper.go:181\n  | runtime.goexit\n  | \t/usr/local/go/src/runtime/asm_amd64.s:1700\nWraps: (2) lock held by current process\nError types: (1) *withstack.withStack (2) *errutil.leafError\ngithub.com/pingcap/errors.Trace\n\t/root/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20250523034308-74f78ae071ee/juju_adaptor.go:15\ngithub.com/pingcap/tidb/pkg/lightning/backend/local.(*engineManager).openEngineDB\n\t/workspace/source/tidb/pkg/lightning/backend/local/engine_mgr.go:230\ngithub.com/pingcap/tidb/pkg/lightning/backend/local.(*engineManager).openEngine\n\t/workspace/source/tidb/pkg/lightning/backend/local/engine_mgr.go:235\ngithub.com/pingcap/tidb/pkg/lightning/backend/local.(*Backend).OpenEngine\n\t/workspace/source/tidb/pkg/lightning/backend/local/local.go:833\ngithub.com/pingcap/tidb/pkg/lightning/backend.EngineManager.OpenEngine\n\t/workspace/source/tidb/pkg/lightning/backend/backend.go:270\ngithub.com/pingcap/tidb/pkg/executor/importer.(*TableImporter).OpenIndexEngine\n\t/workspace/source/tidb/pkg/executor/importer/table_import.go:513\ngithub.com/pingcap/tidb/pkg/disttask/importinto.(*importStepExecutor).RunSubtask\n\t/workspace/source/tidb/pkg/disttask/importinto/task_executor.go:237\ngithub.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.(*BaseTaskExecutor).runSubtask.func1\n\t/workspace/source/tidb/pkg/disttask/framework/taskexecutor/task_executor.go:504\ngithub.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.(*BaseTaskExecutor).runSubtask\n\t/workspace/source/tidb/pkg/disttask/framework/taskexecutor/task_executor.go:505\ngithub.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.(*BaseTaskExecutor).Run\n\t/workspace/source/tidb/pkg/disttask/framework/taskexecutor/task_executor.go:379\ngithub.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.(*Manager).startTaskExecutor.func2\n\t/workspace/source/tidb/pkg/disttask/framework/taskexecutor/manager.go:357\ngithub.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).RunWithLog.func1\n\t/workspace/source/tidb/pkg/util/wait_group_wrapper.go:181\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1700"]
[2025/11/29 22:24:19.012 +08:00] [INFO] [retry.go:80] ["meet un-retryable error"] [error="lock held by current process"] [info="type: *errors.withStack, value: lock held by current process"]
[2025/11/29 22:24:19.013 +08:00] [WARN] [task_executor.go:755] ["subtask failed"] [task-id=120002] [task-key=ImportInto/120002] [error="lock held by current process"] [errorVerbose="lock held by current process\n(1) attached stack trace\n  -- stack trace:\n  | github.com/cockroachdb/pebble/vfs.defaultFS.Lock\n  | \t/root/go/pkg/mod/github.com/cockroachdb/pebble@v1.1.4-0.20250120151818-5dd133a1e6fb/vfs/file_lock_unix.go:50\n  | github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).Lock\n  | \t/root/go/pkg/mod/github.com/cockroachdb/pebble@v1.1.4-0.20250120151818-5dd133a1e6fb/vfs/disk_health.go:679\n  | github.com/cockroachdb/pebble.LockDirectory\n  | \t/root/go/pkg/mod/github.com/cockroachdb/pebble@v1.1.4-0.20250120151818-5dd133a1e6fb/open.go:1068\n  | github.com/cockroachdb/pebble.Open\n  | \t/root/go/pkg/mod/github.com/cockroachdb/pebble@v1.1.4-0.20250120151818-5dd133a1e6fb/open.go:116\n  | github.com/pingcap/tidb/pkg/lightning/backend/local.(*engineManager).openEngineDB\n  | \t/workspace/source/tidb/pkg/lightning/backend/local/engine_mgr.go:229\n  | github.com/pingcap/tidb/pkg/lightning/backend/local.(*engineManager).openEngine\n  | \t/workspace/source/tidb/pkg/lightning/backend/local/engine_mgr.go:235\n  | github.com/pingcap/tidb/pkg/lightning/backend/local.(*Backend).OpenEngine\n  | \t/workspace/source/tidb/pkg/lightning/backend/local/local.go:833\n  | github.com/pingcap/tidb/pkg/lightning/backend.EngineManager.OpenEngine\n  | \t/workspace/source/tidb/pkg/lightning/backend/backend.go:270\n  | github.com/pingcap/tidb/pkg/executor/importer.(*TableImporter).OpenIndexEngine\n  | \t/workspace/source/tidb/pkg/executor/importer/table_import.go:513\n  | github.com/pingcap/tidb/pkg/disttask/importinto.(*importStepExecutor).RunSubtask\n  | \t/workspace/source/tidb/pkg/disttask/importinto/task_executor.go:237\n  | github.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.(*BaseTaskExecutor).runSubtask.func1\n  | \t/workspace/source/tidb/pkg/disttask/framework/taskexecutor/task_executor.go:504\n  | github.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.(*BaseTaskExecutor).runSubtask\n  | \t/workspace/source/tidb/pkg/disttask/framework/taskexecutor/task_executor.go:505\n  | github.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.(*BaseTaskExecutor).Run\n  | \t/workspace/source/tidb/pkg/disttask/framework/taskexecutor/task_executor.go:379\n  | github.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.(*Manager).startTaskExecutor.func2\n  | \t/workspace/source/tidb/pkg/disttask/framework/taskexecutor/manager.go:357\n  | github.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).RunWithLog.func1\n  | \t/workspace/source/tidb/pkg/util/wait_group_wrapper.go:181\n  | runtime.goexit\n  | \t/usr/local/go/src/runtime/asm_amd64.s:1700\nWraps: (2) lock held by current process\nError types: (1) *withstack.withStack (2) *errutil.leafError\ngithub.com/pingcap/errors.Trace\n\t/root/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20250523034308-74f78ae071ee/juju_adaptor.go:15\ngithub.com/pingcap/tidb/pkg/lightning/backend/local.(*engineManager).openEngineDB\n\t/workspace/source/tidb/pkg/lightning/backend/local/engine_mgr.go:230\ngithub.com/pingcap/tidb/pkg/lightning/backend/local.(*engineManager).openEngine\n\t/workspace/source/tidb/pkg/lightning/backend/local/engine_mgr.go:235\ngithub.com/pingcap/tidb/pkg/lightning/backend/local.(*Backend).OpenEngine\n\t/workspace/source/tidb/pkg/lightning/backend/local/local.go:833\ngithub.com/pingcap/tidb/pkg/lightning/backend.EngineManager.OpenEngine\n\t/workspace/source/tidb/pkg/lightning/backend/backend.go:270\ngithub.com/pingcap/tidb/pkg/executor/importer.(*TableImporter).OpenIndexEngine\n\t/workspace/source/tidb/pkg/executor/importer/table_import.go:513\ngithub.com/pingcap/tidb/pkg/disttask/importinto.(*importStepExecutor).RunSubtask\n\t/workspace/source/tidb/pkg/disttask/importinto/task_executor.go:237\ngithub.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.(*BaseTaskExecutor).runSubtask.func1\n\t/workspace/source/tidb/pkg/disttask/framework/taskexecutor/task_executor.go:504\ngithub.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.(*BaseTaskExecutor).runSubtask\n\t/workspace/source/tidb/pkg/disttask/framework/taskexecutor/task_executor.go:505\ngithub.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.(*BaseTaskExecutor).Run\n\t/workspace/source/tidb/pkg/disttask/framework/taskexecutor/task_executor.go:379\ngithub.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.(*Manager).startTaskExecutor.func2\n\t/workspace/source/tidb/pkg/disttask/framework/taskexecutor/manager.go:357\ngithub.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).RunWithLog.func1\n\t/workspace/source/tidb/pkg/util/wait_group_wrapper.go:181\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1700"]
[2025/11/29 22:24:19.022 +08:00] [ERROR] [task_executor.go:384] ["run subtask failed"] [task-id=120002] [task-key=ImportInto/120002] [error="lock held by current process"] [errorVerbose="lock held by current process\n(1) attached stack trace\n  -- stack trace:\n  | github.com/cockroachdb/pebble/vfs.defaultFS.Lock\n  | \t/root/go/pkg/mod/github.com/cockroachdb/pebble@v1.1.4-0.20250120151818-5dd133a1e6fb/vfs/file_lock_unix.go:50\n  | github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).Lock\n  | \t/root/go/pkg/mod/github.com/cockroachdb/pebble@v1.1.4-0.20250120151818-5dd133a1e6fb/vfs/disk_health.go:679\n  | github.com/cockroachdb/pebble.LockDirectory\n  | \t/root/go/pkg/mod/github.com/cockroachdb/pebble@v1.1.4-0.20250120151818-5dd133a1e6fb/open.go:1068\n  | github.com/cockroachdb/pebble.Open\n  | \t/root/go/pkg/mod/github.com/cockroachdb/pebble@v1.1.4-0.20250120151818-5dd133a1e6fb/open.go:116\n  | github.com/pingcap/tidb/pkg/lightning/backend/local.(*engineManager).openEngineDB\n  | \t/workspace/source/tidb/pkg/lightning/backend/local/engine_mgr.go:229\n  | github.com/pingcap/tidb/pkg/lightning/backend/local.(*engineManager).openEngine\n  | \t/workspace/source/tidb/pkg/lightning/backend/local/engine_mgr.go:235\n  | github.com/pingcap/tidb/pkg/lightning/backend/local.(*Backend).OpenEngine\n  | \t/workspace/source/tidb/pkg/lightning/backend/local/local.go:833\n  | github.com/pingcap/tidb/pkg/lightning/backend.EngineManager.OpenEngine\n  | \t/workspace/source/tidb/pkg/lightning/backend/backend.go:270\n  | github.com/pingcap/tidb/pkg/executor/importer.(*TableImporter).OpenIndexEngine\n  | \t/workspace/source/tidb/pkg/executor/importer/table_import.go:513\n  | github.com/pingcap/tidb/pkg/disttask/importinto.(*importStepExecutor).RunSubtask\n  | \t/workspace/source/tidb/pkg/disttask/importinto/task_executor.go:237\n  | github.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.(*BaseTaskExecutor).runSubtask.func1\n  | \t/workspace/source/tidb/pkg/disttask/framework/taskexecutor/task_executor.go:504\n  | github.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.(*BaseTaskExecutor).runSubtask\n  | \t/workspace/source/tidb/pkg/disttask/framework/taskexecutor/task_executor.go:505\n  | github.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.(*BaseTaskExecutor).Run\n  | \t/workspace/source/tidb/pkg/disttask/framework/taskexecutor/task_executor.go:379\n  | github.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.(*Manager).startTaskExecutor.func2\n  | \t/workspace/source/tidb/pkg/disttask/framework/taskexecutor/manager.go:357\n  | github.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).RunWithLog.func1\n  | \t/workspace/source/tidb/pkg/util/wait_group_wrapper.go:181\n  | runtime.goexit\n  | \t/usr/local/go/src/runtime/asm_amd64.s:1700\nWraps: (2) lock held by current process\nError types: (1) *withstack.withStack (2) *errutil.leafError\ngithub.com/pingcap/errors.Trace\n\t/root/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20250523034308-74f78ae071ee/juju_adaptor.go:15\ngithub.com/pingcap/tidb/pkg/lightning/backend/local.(*engineManager).openEngineDB\n\t/workspace/source/tidb/pkg/lightning/backend/local/engine_mgr.go:230\ngithub.com/pingcap/tidb/pkg/lightning/backend/local.(*engineManager).openEngine\n\t/workspace/source/tidb/pkg/lightning/backend/local/engine_mgr.go:235\ngithub.com/pingcap/tidb/pkg/lightning/backend/local.(*Backend).OpenEngine\n\t/workspace/source/tidb/pkg/lightning/backend/local/local.go:833\ngithub.com/pingcap/tidb/pkg/lightning/backend.EngineManager.OpenEngine\n\t/workspace/source/tidb/pkg/lightning/backend/backend.go:270\ngithub.com/pingcap/tidb/pkg/executor/importer.(*TableImporter).OpenIndexEngine\n\t/workspace/source/tidb/pkg/executor/importer/table_import.go:513\ngithub.com/pingcap/tidb/pkg/disttask/importinto.(*importStepExecutor).RunSubtask\n\t/workspace/source/tidb/pkg/disttask/importinto/task_executor.go:237\ngithub.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.(*BaseTaskExecutor).runSubtask.func1\n\t/workspace/source/tidb/pkg/disttask/framework/taskexecutor/task_executor.go:504\ngithub.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.(*BaseTaskExecutor).runSubtask\n\t/workspace/source/tidb/pkg/disttask/framework/taskexecutor/task_executor.go:505\ngithub.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.(*BaseTaskExecutor).Run\n\t/workspace/source/tidb/pkg/disttask/framework/taskexecutor/task_executor.go:379\ngithub.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.(*Manager).startTaskExecutor.func2\n\t/workspace/source/tidb/pkg/disttask/framework/taskexecutor/manager.go:357\ngithub.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).RunWithLog.func1\n\t/workspace/source/tidb/pkg/util/wait_group_wrapper.go:181\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1700"]
```
### 2. What did you expect to see? (Required)


### 3. What did you see instead (Required)

### 4. What is your TiDB version? (Required)

--- a/pkg/dxf/importinto/task_executor.go
+++ b/pkg/dxf/importinto/task_executor.go
@@ -258,6 +258,15 @@ func (s *importStepExecutor) RunSubtask(ctx context.Context, subtask *proto.Subt
 		indexKVFileCount: &indexKVFiles,
 	}
 	s.sharedVars.Store(subtaskMeta.ID, sharedVars)
+	defer func() {
+		s.sharedVars.Delete(subtaskMeta.ID)
+		if err == nil || !s.tableImporter.IsLocalSort() {
+			return
+		}
+		if cleanupErr := s.tableImporter.Backend().CleanupAllLocalEngines(context.Background()); cleanupErr != nil {
+			logger.Warn("cleanup engines failed", zap.Error(cleanupErr))
+		}
+	}()
 
 	wctx := workerpool.NewContext(ctx)
 	tasks := make([]*importStepMinimalTask, 0, len(subtaskMeta.Chunks))

--- a/pkg/dxf/importinto/task_executor.go
+++ b/pkg/dxf/importinto/task_executor.go
@@ -367,7 +367,6 @@ func (s *importStepExecutor) onFinished(ctx context.Context, subtask *proto.Subt
 		}
 	}
 
-	s.sharedVars.Delete(subtaskMeta.ID)
 	newMeta, err := subtaskMeta.Marshal()
 	if err != nil {
 		return errors.Trace(err)

--- a/pkg/executor/importer/table_import.go
+++ b/pkg/executor/importer/table_import.go
@@ -549,6 +549,7 @@ func (ti *TableImporter) OpenDataEngine(ctx context.Context, engineID int32) (*b
 func (ti *TableImporter) ImportAndCleanup(ctx context.Context, closedEngine *backend.ClosedEngine) (int64, error) {
 	var kvCount int64
 	importErr := closedEngine.Import(ctx, ti.regionSplitSize, ti.regionSplitKeys)
+	failpoint.InjectCall("mockDataEngineImportErr", &importErr)
 	if common.ErrFoundDuplicateKeys.Equal(importErr) {
 		importErr = local.ConvertToErrFoundConflictRecords(importErr, ti.encTable)
 	}

--- a/pkg/lightning/backend/local/engine.go
+++ b/pkg/lightning/backend/local/engine.go
@@ -194,6 +194,32 @@ func (e *Engine) Exist(dataDir string) error {
 	return nil
 }
 
+// finishWrite flushes all data and waits for the background routine to finish.
+// after calling this method, the engine cannot be written anymore. the underlying
+// pebble DB is still kept open for reading.
+func (e *Engine) finishWrite(ctx context.Context) error {
+	e.rLock()
+	if e.closed.Load() {
+		e.rUnlock()
+		return nil
+	}
+
+	err := e.flushEngineWithoutLock(ctx)
+	e.rUnlock()
+
+	// use mutex to make sure we won't close sstMetasChan while other routines
+	// trying to do flush.
+	e.lock(importMutexStateClose)
+	e.closed.Store(true)
+	close(e.sstMetasChan)
+	e.unlock()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	e.wg.Wait()
+	return e.ingestErr.Get()
+}
+
 func isStateLocked(state importMutexState) bool {
 	return state&(importMutexStateClose|importMutexStateImport) != 0
 }

--- a/pkg/lightning/backend/local/engine_mgr.go
+++ b/pkg/lightning/backend/local/engine_mgr.go
@@ -40,6 +40,7 @@ import (
 	"github.com/tikv/client-go/v2/oracle"
 	tikvclient "github.com/tikv/client-go/v2/tikv"
 	"go.uber.org/atomic"
+	"go.uber.org/multierr"
 	"go.uber.org/zap"
 	"golang.org/x/sync/errgroup"
 )
@@ -351,26 +352,7 @@ func (em *engineManager) closeEngine(
 	}
 
 	engine := engineI.(*Engine)
-	engine.rLock()
-	if engine.closed.Load() {
-		engine.rUnlock()
-		return nil
-	}
-
-	err := engine.flushEngineWithoutLock(ctx)
-	engine.rUnlock()
-
-	// use mutex to make sure we won't close sstMetasChan while other routines
-	// trying to do flush.
-	engine.lock(importMutexStateClose)
-	engine.closed.Store(true)
-	close(engine.sstMetasChan)
-	engine.unlock()
-	if err != nil {
-		return errors.Trace(err)
-	}
-	engine.wg.Wait()
-	return engine.ingestErr.Get()
+	return engine.finishWrite(ctx)
 }
 
 // getImportedKVCount returns the number of imported KV pairs of some engine.
@@ -492,6 +474,27 @@ func (em *engineManager) cleanupEngine(ctx context.Context, engineUUID uuid.UUID
 	localEngine.TotalSize.Store(0)
 	localEngine.Length.Store(0)
 	return nil
+}
+
+// cleanupAllLocalEngines closes and removes all local engines, used for best-effort cleanup on error.
+func (em *engineManager) cleanupAllLocalEngines(ctx context.Context) error {
+	var (
+		retErr  error
+		engines []*Engine
+	)
+	em.engines.Range(func(_, v any) bool {
+		engines = append(engines, v.(*Engine))
+		return true
+	})
+	for _, eng := range engines {
+		if err := eng.finishWrite(ctx); err != nil {
+			retErr = multierr.Append(retErr, err)
+		}
+		if err := em.cleanupEngine(ctx, eng.UUID); err != nil {
+			retErr = multierr.Append(retErr, err)
+		}
+	}
+	return retErr
 }
 
 // LocalWriter returns a new local writer.

--- a/pkg/lightning/backend/local/local.go
+++ b/pkg/lightning/backend/local/local.go
@@ -836,6 +836,11 @@ func (local *Backend) FlushAllEngines(parentCtx context.Context) (err error) {
 	return local.engineMgr.flushAllEngines(parentCtx)
 }
 
+// CleanupAllLocalEngines closes and removes all local engines, used for best-effort cleanup on error.
+func (local *Backend) CleanupAllLocalEngines(ctx context.Context) error {
+	return local.engineMgr.cleanupAllLocalEngines(ctx)
+}
+
 // RetryImportDelay returns the delay time before retrying to import a file.
 func (*Backend) RetryImportDelay() time.Duration {
 	return defaultRetryBackoffTime

--- a/tests/realtikvtest/importintotest3/BUILD.bazel
+++ b/tests/realtikvtest/importintotest3/BUILD.bazel
@@ -4,6 +4,7 @@ go_test(
     name = "importintotest3_test",
     timeout = "moderate",
     srcs = [
+        "cleanup_on_error_test.go",
         "cross_ks_test.go",
         "extra_param_test.go",
         "file_compression_test.go",

--- a/tests/realtikvtest/importintotest3/BUILD.bazel
+++ b/tests/realtikvtest/importintotest3/BUILD.bazel
@@ -28,6 +28,7 @@ go_test(
         "//pkg/planner/core",
         "//pkg/sessionctx/vardef",
         "//pkg/store",
+        "//pkg/store/driver/error",
         "//pkg/testkit",
         "//pkg/testkit/testfailpoint",
         "//tests/realtikvtest",

--- a/tests/realtikvtest/importintotest3/cleanup_on_error_test.go
+++ b/tests/realtikvtest/importintotest3/cleanup_on_error_test.go
@@ -41,7 +41,7 @@ func (s *mockGCSSuite) TestCleanupOnDataEngineImportError() {
 		}
 	})
 	s.tk.MustQuery(fmt.Sprintf("IMPORT INTO t FROM '%s'", dataPath))
-	s.GreaterOrEqual(enterCnt.Load(), 2, "the import should retry on data engine import error")
+	s.GreaterOrEqual(enterCnt.Load(), int32(2), "the import should retry on data engine import error")
 
 	s.tk.MustQuery("select * from t;").Sort().Check(testkit.Rows("1 a", "2 b"))
 }

--- a/tests/realtikvtest/importintotest3/cleanup_on_error_test.go
+++ b/tests/realtikvtest/importintotest3/cleanup_on_error_test.go
@@ -1,0 +1,47 @@
+// Copyright 2026 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package importintotest
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync/atomic"
+
+	drivererr "github.com/pingcap/tidb/pkg/store/driver/error"
+	"github.com/pingcap/tidb/pkg/testkit"
+	"github.com/pingcap/tidb/pkg/testkit/testfailpoint"
+)
+
+func (s *mockGCSSuite) TestCleanupOnDataEngineImportError() {
+	s.prepareAndUseDB("cleanup_on_error")
+	s.tk.MustExec("create table t (a int primary key, b varchar(32));")
+
+	tempDir := s.T().TempDir()
+	dataPath := filepath.Join(tempDir, "data.csv")
+	content := []byte("1,a\n2,b\n")
+	s.NoError(os.WriteFile(dataPath, content, 0o644))
+
+	var enterCnt atomic.Int32
+	testfailpoint.EnableCall(s.T(), "github.com/pingcap/tidb/pkg/executor/importer/mockDataEngineImportErr", func(errP *error) {
+		if enterCnt.Add(1) == 1 {
+			*errP = drivererr.ErrPDServerTimeout
+		}
+	})
+	s.tk.MustQuery(fmt.Sprintf("IMPORT INTO t FROM '%s'", dataPath))
+	s.GreaterOrEqual(enterCnt.Load(), 2, "the import should retry on data engine import error")
+
+	s.tk.MustQuery("select * from t;").Sort().Check(testkit.Rows("1 a", "2 b"))
+}

